experimental/yarpl/include/reactivestreams/ReactiveStreams.h

```diff
@@ -0,0 +1,39 @@
+// Copyright 2004-present Facebook. All Rights Reserved.
+
+#pragma once
+
+#include <exception>
+#include <memory>
+
+namespace reactivestreams_yarpl {
+// TODO remove 'yarpl' if we agree on this usage elsewhere
+
+class Subscription {
+ public:
+  virtual ~Subscription() = default;
+  virtual void request(int64_t n) = 0;
+  virtual void cancel() = 0;
+};
+
+template <typename T>
+class Subscriber {
+ public:
+  virtual ~Subscriber() = default;
+  virtual void onSubscribe(Subscription*) = 0;
```

Comment:
Documented as:

```
  /**
   * Receive the Subscription that controls the lifecycle of this
   * Subscriber/Subscription pair.
   *
   * This Subscription* will exist until onComplete, onError, or cancellation.
   *
   * Rules:
   * - Do NOT use Subscription* after onComplete();
   * - Do NOT use Subscription* after onError();
   * - Do NOT use Subscription* after calling Subscription*->cancel();
   */
  virtual void onSubscribe(Subscription*) = 0;
```

Having been working with this for a few days, it actually is pretty nice working with this knowing that only the `Subscription` concerns itself with lifecycle.